### ISSUE TITLE
Ensuring order in case of exception in System.Threading.Tasks.Dataflow blocks

### ISSUE THREAD

**Issue by giacomociti** [dotnet/corefx#28033]
Dataflow blocks shut down in case of an exception, but a few more items may get propagated before actual completion, including items that appeared after the faulty one in the input sequence.
I may provide a few lines of code to show this, but looking at the source code it seems this is intended. I don't get the rationale of this behavior: skipping a faulty item but propagating a few subsequent ones seems at odd with the (very useful) ordering guarantee that all blocks provide by default.

In my use case, for example, I process items (file lines) with some degree of parallelism in a transform block, linked to a (sequential) action block in charge of storing a kind of 'commit point'.
In case of an error I'd like to restart from the previously saved commit point, skipping most of the input lines already processed but still ensuring 'at least once' processing.

I think that items following a faulty one should not be propagated.

**Comment by danmosemsft** [comment]
@karelz do we have a designated owner for Dataflow?

**Comment by karelz** [comment]
It is part of System.Threading which is [owned](https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/issue-guide.md#areas) by @kouvel
TPL part was always owned by someone on BCL team - last was @tarekgh. @tarekgh was there any transition to someone else?

**Comment by tarekgh** [comment]
@giacomociti could you provide some sample code demonstrate the issue?

@karelz I own System.Linq.Parallel and we had implicit understanding System.Threading.Tasks owned by @kouvel and I was helping with issues in this area. 

CC @stephentoub

**Comment by giacomociti** [comment]
this is a [gist](https://gist.github.com/giacomociti/1e95d95d6a0d02d60472e9207d8c2b9c) with an example.
Thanks

**Comment by tarekgh** [comment]
@giacomociti I'll take a look and get back to you. thank you.

**Comment by tarekgh** [comment]
Thanks @giacomociti for your patience. Here is what is going on and why you see more items processed even after throwing the exception:

In your Run method:

```C#
        private static async Task Run<T>(ITargetBlock<T> firstBlock, IDataflowBlock lastBlock, IEnumerable<T> input)
        {
            foreach (var item in input)
            {
                var sent = await firstBlock.SendAsync(item);
                if (!sent) break;
            }
            firstBlock.Complete();
            await lastBlock.Completion;
        }
```

If you look at the doc of TransformBlock.SendAsync https://msdn.microsoft.com/en-us/library/hh194681(v=vs.110).aspx it says 

```
Asynchronously offers a message to the target message block, allowing for postponement.

Return Value
Type: System.Threading.Tasks.Task<Boolean>
A Task<TResult> that represents the asynchronous send. If the target accepts and consumes the offered element during the call to SendAsync, upon return from the call the resulting Task<TResult> will be completed and its Result property will return true. If the target declines the offered element during the call, upon return from the call the resulting Task<TResult> will be completed and its Result property will return false. If the target postpones the offered element, the element will be buffered until such time that the target consumes or releases it, at which point the task will complete, with its Result indicating whether the message was consumed. If the target never attempts to consume or release the message, the returned task will never complete.
```

That means SendAsync is just posting the request to the target and this request will get processed later asynchronously if it couldn't get processed quickly. when returning from the await of the SendAsync, means the request is queued and not necessary processed. here is a stack that can help show how the SendAsync posts the request and end up scheduling it for processing asynchronously:

```
	System.Threading.Tasks.Dataflow.dll!System.Threading.Tasks.Dataflow.Internal.TargetCore<int>.ProcessAsyncIfNecessary_Slow(bool repeat) Line 379	C#
 	System.Threading.Tasks.Dataflow.dll!System.Threading.Tasks.Dataflow.Internal.TargetCore<int>.ProcessAsyncIfNecessary(bool repeat) Line 350	C#
 	System.Threading.Tasks.Dataflow.dll!System.Threading.Tasks.Dataflow.Internal.TargetCore<int>.OfferMessage(System.Threading.Tasks.Dataflow.DataflowMessageHeader messageHeader, int messageValue, System.Threading.Tasks.Dataflow.ISourceBlock<int> source, bool consumeToAccept) Line 227	C#
 	System.Threading.Tasks.Dataflow.dll!System.Threading.Tasks.Dataflow.TransformBlock<int, int>.System.Threading.Tasks.Dataflow.ITargetBlock<int>.OfferMessage(System.Threading.Tasks.Dataflow.DataflowMessageHeader messageHeader, int messageValue, System.Threading.Tasks.Dataflow.ISourceBlock<int> source, bool consumeToAccept) Line 374	C#
 	System.Threading.Tasks.Dataflow.dll!System.Threading.Tasks.Dataflow.DataflowBlock.SendAsync<int>(System.Threading.Tasks.Dataflow.ITargetBlock<int> target, int item, System.Threading.CancellationToken cancellationToken) Line 315	C#
 	System.Threading.Tasks.Dataflow.dll!System.Threading.Tasks.Dataflow.DataflowBlock.SendAsync<int>(System.Threading.Tasks.Dataflow.ITargetBlock<int> target, int item) Line 273	C#
 	MyTest.Tests.dll!System.Threading.Tasks.Dateflow.Tests.DateflowTest.Run<int>(System.Threading.Tasks.Dataflow.ITargetBlock<int> firstBlock, System.Threading.Tasks.Dataflow.IDataflowBlock lastBlock, System.Collections.Generic.IEnumerable<int> input) Line 52	C#

```

Later when the asynchronously scheduled request get processed which will call your delegate 

```C#
 x => x == badItem? throw new Exception("bug") : x
```

```
	MyTest.Tests.dll!System.Threading.Tasks.Dateflow.Tests.DateflowTest.TransformBlockPreserveOrder.AnonymousMethod__0_0(int x) Line 31	C#
 	System.Threading.Tasks.Dataflow.dll!System.Threading.Tasks.Dataflow.TransformBlock<int, int>.ProcessMessage(System.Func<int, int> transform, System.Collections.Generic.KeyValuePair<int, long> messageWithId) Line 183	C#
 	System.Threading.Tasks.Dataflow.dll!System.Threading.Tasks.Dataflow.TransformBlock<int, int>..ctor.AnonymousMethod__3(System.Collections.Generic.KeyValuePair<int, long> messageWithId) Line 127	C#
 	System.Threading.Tasks.Dataflow.dll!System.Threading.Tasks.Dataflow.Internal.TargetCore<int>.ProcessMessagesLoopCore() Line 489	C#
 	System.Threading.Tasks.Dataflow.dll!System.Threading.Tasks.Dataflow.Internal.TargetCore<int>.ProcessAsyncIfNecessary_Slow.AnonymousMethod__34_0(object thisTargetCore) Line 379	C#
 	System.Private.CoreLib.dll!System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state) Line 167	C#
```

If your delegate throws an exception, it will stop queuing any more requests but it will try to process all already queued requests before shutting down the Dataflow block and that is why you see more items processed even after throwing the exception.

In summary, the current design process all queued requests before shutting down the Dataflow block. I agree someone else can argue we should drop all queued messages which ordered after the failed one, but this is not what current design does.

Let me know if you have any more questions

**Comment by tarekgh** [comment]
I can see it either way. if SendAsync succeeded to queue item, it would be surprising too not processing it even if there is one of the queued items caused a fault. anyway, this is the option picked by the design. thanks for reporting this.

**Comment by giacomociti** [comment]
Processing items already queued is ok, the issue is propagating their result. Anyway I appreciate your analysis, thanks again.
